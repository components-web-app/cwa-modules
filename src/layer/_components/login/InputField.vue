<template>
  <div>
    <label :for="id" class="cwa-block cwa-text-sm cwa-font-medium cwa-leading-6 cwa-text-neutral-300">{{ props.label }}</label>
    <div class="cwa-mt-2">
      <input
        :id="id"
        :name="props.name"
        :type="props.type"
        :autocomplete="props.autocomplete"
        :required="props.required"
        :value="modelValue"
        class="cwa-block cwa-w-full cwa-border-0 cwa-py-1.5 cwa-text-neutral-900 cwa-shadow-sm cwa-ring-1 cwa-ring-inset cwa-ring-neutral-300 placeholder:cwa-text-neutral-400 focus:cwa-ring-2 focus:cwa-ring-inset focus:cwa-ring-neutral-600 sm:cwa-text-sm sm:cwa-leading-6"
        @input="$emit('update:modelValue', $event.target.value)"
      >
      <div v-if="errors">
        <CwaUiAlertWarning v-for="(error, index) in errors" :key="`error-${id}-${index}`" class="cwa-mt-2">
          {{ error }}
        </CwaUiAlertWarning>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { v4 as uuidv4 } from 'uuid'
import { ref } from 'vue'

defineEmits(['update:modelValue'])

const props = withDefaults(defineProps<{
  name: string,
  type?: string,
  label: string,
  autocomplete?: string,
  required?: boolean,
  modelValue: string,
  errors?: string[]
}>(), {
  type: 'text',
  autocomplete: undefined,
  required: false,
  modelValue: '',
  errors: undefined
})

const id = ref(uuidv4())
</script>
