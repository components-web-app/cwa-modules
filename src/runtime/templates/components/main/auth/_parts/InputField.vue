<template>
  <div>
    <label
      :for="id"
      class="cwa-block cwa-text-sm cwa-font-medium cwa-leading-6 cwa-text-neutral-300"
    >{{ props.label }}</label>
    <div class="cwa-mt-2">
      <input
        :id="id"
        :name="props.name"
        :type="props.type"
        :autocomplete="props.autocomplete"
        :required="props.required"
        :value="modelValue"
        class="cwa-block cwa-w-full cwa-border-0 cwa-py-1.5 cwa-text-neutral-900 cwa-shadow-sm cwa-ring-1 cwa-ring-inset cwa-ring-neutral-300 placeholder:cwa-text-neutral-400 focus:cwa-ring-2 focus:cwa-ring-inset focus:cwa-ring-neutral-600 sm:cwa-text-sm sm:cwa-leading-6"
        @input="updateModelValue"
      >
      <div v-if="errors">
        <CwaUiAlertWarning
          v-for="(error, index) in errors"
          :key="`error-${id}-${index}`"
          class="cwa-mt-2"
        >
          {{ error }}
        </CwaUiAlertWarning>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { useId } from '#app'

const emit = defineEmits(['update:modelValue'])

const props = withDefaults(defineProps<{
  name: string
  type?: string
  label: string
  autocomplete?: string
  required?: boolean
  modelValue: string
  errors?: string[]
}>(), {
  type: 'text',
  autocomplete: undefined,
  required: false,
  modelValue: '',
  errors: undefined,
})

const id = useId()

function updateModelValue($event: Event) {
  if (!($event.target instanceof HTMLInputElement)) {
    return
  }
  emit('update:modelValue', $event.target?.value)
}
</script>
