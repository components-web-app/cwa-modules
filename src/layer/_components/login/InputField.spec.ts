// @vitest-environment nuxt
import { describe, expect, test, vi } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import InputField from './InputField.vue'

vi.mock('uuid', () => ({
  v4 () {
    return 'this-is-unique-string'
  }
}))

interface IInputMeta {
  name: string;
  label: string;
  modelValue: string;
  type?: string;
  autocomplete?: string;
  required?: boolean;
  errors?: string[];
}

function createWrapper ({
  name,
  label,
  modelValue,
  type = 'text',
  autocomplete = 'login',
  required = false,
  errors
}: IInputMeta) {
  return shallowMount(InputField, {
    props: {
      name,
      label,
      modelValue,
      type,
      autocomplete,
      required,
      errors
    },
    global: {
      renderStubDefaultSlot: true
    }
  })
}

describe('InputField', () => {
  describe('emits', () => {
    test('should emit updated value', async () => {
      const mockNewValue = 'new input value'
      const wrapper = createWrapper({
        name: 'login',
        label: 'Login',
        modelValue: ''
      })

      const input = wrapper.find('input')

      await input.setValue(mockNewValue)

      expect(wrapper.emitted('update:modelValue')).toEqual([[mockNewValue]])
    })
  })

  describe('snapshots', () => {
    test('should match snapshot with passed content', () => {
      const wrapper = createWrapper({
        name: 'login',
        label: 'Login',
        modelValue: ''
      })

      expect(wrapper.element).toMatchSnapshot()
    })

    test('should match snapshot IF errors are present', () => {
      const wrapper = createWrapper({
        name: 'login',
        label: 'Login',
        modelValue: '',
        errors: ['Login should be minimum 6 characters long', 'Login cannot contain whitespaces']
      })

      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
