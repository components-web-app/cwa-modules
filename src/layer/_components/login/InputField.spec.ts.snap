// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputField > snapshots > should match snapshot IF errors are present 1`] = `
<div>
  <label
    class="cwa-block cwa-text-sm cwa-font-medium cwa-leading-6 cwa-text-neutral-300"
    for="this-is-unique-string"
  >
    Login
  </label>
  <div
    class="cwa-mt-2"
  >
    <input
      autocomplete="login"
      class="cwa-block cwa-w-full cwa-rounded-md cwa-border-0 cwa-py-1.5 cwa-text-neutral-900 cwa-shadow-sm cwa-ring-1 cwa-ring-inset cwa-ring-neutral-300 placeholder:cwa-text-neutral-400 focus:cwa-ring-2 focus:cwa-ring-inset focus:cwa-ring-neutral-600 sm:cwa-text-sm sm:cwa-leading-6"
      id="this-is-unique-string"
      name="login"
      type="text"
    />
    <div>
      
      <cwa-utils-alert-warning-stub
        class="cwa-mt-2"
      >
        Login should be minimum 6 characters long
      </cwa-utils-alert-warning-stub>
      <cwa-utils-alert-warning-stub
        class="cwa-mt-2"
      >
        Login cannot contain whitespaces
      </cwa-utils-alert-warning-stub>
      
    </div>
  </div>
</div>
`;

exports[`InputField > snapshots > should match snapshot with passed content 1`] = `
<div>
  <label
    class="cwa-block cwa-text-sm cwa-font-medium cwa-leading-6 cwa-text-neutral-300"
    for="this-is-unique-string"
  >
    Login
  </label>
  <div
    class="cwa-mt-2"
  >
    <input
      autocomplete="login"
      class="cwa-block cwa-w-full cwa-rounded-md cwa-border-0 cwa-py-1.5 cwa-text-neutral-900 cwa-shadow-sm cwa-ring-1 cwa-ring-inset cwa-ring-neutral-300 placeholder:cwa-text-neutral-400 focus:cwa-ring-2 focus:cwa-ring-inset focus:cwa-ring-neutral-600 sm:cwa-text-sm sm:cwa-leading-6"
      id="this-is-unique-string"
      name="login"
      type="text"
    />
    <!--v-if-->
  </div>
</div>
`;
